### PR TITLE
USDT tracing - net

### PR DESCRIPTION
**Pull request by Naviabheeman**
This PR ports USDT support for inbound and outbound network messages to tapyrus.

### CODE REVIEW

File: doc/dependencies.md [hunk 1]
```diff
@@ -22,6 +22,7 @@ These are the dependencies currently used by Bitcoin Core. You can find instruct
 | Python (tests) |  | [3.4](https://www.python.org/downloads) |  |  |  |
 | qrencode | [3.4.4](https://fukuchi.org/works/qrencode) |  | No |  |  |
 | Qt | [5.9.6](https://download.qt.io/official_releases/qt/) | 5.x | No |  |  |
+| systemtap | [4.7](https://sourceware.org/systemtap/) |  |  |  |  |
```
Comment (azuchi) [inline doc/dependencies.md:25]: Bitcoin does not seem to define a specific minimum version.

```suggestion
| [systemtap](https://sourceware.org/systemtap/) ([tracing](tracing.md))| N/A |
```

File: doc/tracing.md [hunk 2]
```diff
@@ -0,0 +1,244 @@
+# User-space, Statically Defined Tracing (USDT) for Bitcoin Core
+
+Bitcoin Core includes statically defined tracepoints to allow for more
```
Comment (azuchi) [inline doc/tracing.md:3]:
```suggestion
Tapyrus Core includes statically defined tracepoints to allow for more
```

Other similar modifications are needed.

File: contrib/tracing/log_raw_p2p_msgs.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        except KeyboardInterrupt:
+            exit()
+
+
+if __name__ == "__main__":
+    if len(sys.argv) < 2:
+        print("USAGE:", sys.argv[0], "path/to/bitcoind")
+        exit()
+    path = sys.argv[1]
+    main(path)
```
Comment (azuchi) [inline contrib/tracing/log_raw_p2p_msgs.py:180]: Rename bitcoin to tapyrus.

File: contrib/tracing/p2p_monitor.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+                        i + 3, 1, " %s (%d byte) --->" %
+                        (msg.msg_type, msg.size), curses.A_NORMAL)
+
+
+if __name__ == "__main__":
+    if len(sys.argv) < 2:
+        print("USAGE:", sys.argv[0], "path/to/bitcoind")
+        exit()
+    path = sys.argv[1]
+    main(path)
```
Comment (azuchi) [inline contrib/tracing/p2p_monitor.py:250]: Rename bitcoin to tapyrus

File: doc/build-unix.md [hunk 5]
```diff
@@ -92,6 +92,10 @@ ZMQ dependencies (provides ZMQ API 4.x):
```
Comment (azuchi) [inline doc/build-unix.md:92]: Add the following to the table Optional dependencies.

```
 systemtap   | Tracing (USDT)   | Optional, statically defined tracepoints
```

File: src/net.h [hunk 6]
```diff
@@ -21,6 +21,7 @@
 #include <sync.h>
 #include <uint256.h>
 #include <threadinterrupt.h>
+#include <ui_interface.h>
```
Comment (azuchi) [inline src/net.h:24]: This change need?

File: test/functional/test_runner.py [hunk 7]
```diff
@@ -224,8 +224,12 @@
     'rpc_scantxoutset.py'
 ]
 
+USDT_SCRIPTS = [
+    'interface_usdt_net.py',
+]
+
 # Place EXTENDED_SCRIPTS first since it has the 3 longest running tests
-ALL_SCRIPTS = EXTENDED_SCRIPTS + BASE_SCRIPTS
+ALL_SCRIPTS = EXTENDED_SCRIPTS + BASE_SCRIPTS + USDT_SCRIPTS
```
Comment (azuchi) [inline test/functional/test_runner.py:232]: In Bitcoin, `interface_usdt_net.py` seems to be added `BASE_SCRIPTS`.

**Review by azuchi — CHANGES_REQUESTED** [review]
Bitcoin has also changed the following files:

* depends/config.site.in
* depends/Makelfie

File: contrib/tracing/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,75 @@
+#for kernel libraries
+FROM linuxkit/kernel:4.15 as ksrc
+
```
[the remark this answers is not in the document]
Comment (Naviabheeman) [reply contrib/tracing/Dockerfile:3]: Since we can't test these tracepoints on Macos, I created this container with tapyrus and usdt packages. I use this container to build and test the tracepoints. visual studio code docker integration makes coding and debugging inside the container so convenient. It need not be in the repository if I'm the only person using it though.

File: test/functional/test_runner.py (hunk 7 above)
Comment (Naviabheeman) [reply test/functional/test_runner.py:232]: Yes, bitcoin has support in the test framework to check if a module is present and skip tests if it is missing. I am disabling these tests as they do not complete successfully.

File: doc/dependencies.md (hunk 1 above)
Comment (Naviabheeman) [reply doc/dependencies.md:25]: This is our "version used" column. Bitcoin's table has been simplified a lot. Ours still has "version used" and "minimum version" separately.

File: contrib/tracing/Dockerfile (hunk 8 above)
[the remark this answers is not in the document]
Comment (azuchi) [reply contrib/tracing/Dockerfile:3]: If you change `Dockerfile` in the root dir, wouldn't this be necessary?

File: src/net.h (hunk 6 above)
Comment (Naviabheeman) [reply src/net.h]: no. I'll remove it

File: contrib/tracing/Dockerfile (hunk 8 above)
[the remark this answers is not in the document]
Comment (Naviabheeman) [reply contrib/tracing/Dockerfile]: Yes, this works! I was able to combine the two files. Now build, test and trace are all available inside the tapyrus container.

[the remark this answers is not in the document]
Comment (Naviabheeman) [reply contrib/tracing/Dockerfile]: I verified that the container can build successfully. But didn't check if bpftrace or bcc can be run inside it.

File: Dockerfile [hunk 9]
```diff
@@ -1,8 +1,16 @@
+#for kernel libraries
+FROM linuxkit/kernel:4.15 as ksrc
```
Comment (azuchi) [inline Dockerfile:2]: Why is this necessary? Doesn't it work if you install bpftrace with apt?

**Review by azuchi — CHANGES_REQUESTED** [review]

File: contrib/tracing/log_raw_p2p_msgs.py [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+              event.peer_conn_type.decode("utf-8"),
+              event.peer_addr.decode("utf-8"),
+              event.msg_size,
+              bytes(event.msg[:event.msg_size]).hex(),
+          )
+          )
+
+
+def main(tapyrusd_path):
+    bitcoind_with_usdts = USDT(path=str(tapyrusd_path))
```
Comment (azuchi) [inline contrib/tracing/log_raw_p2p_msgs.py:133]: You can rename `bitcoind_with_usdts` to `tapyrusd_with_usdts`

**Review by azuchi — CHANGES_REQUESTED** [review]

**Review by azuchi — CHANGES_REQUESTED** [review]

File: contrib/tracing/README.md [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+   11  1156     253570 byte  3431    2394924 byte  outbound-full-relay    XXX.X.XX.179:8333
+   13  3425    1809620 byte  1236     305458 byte  inbound                XXX.X.X.X:60380
+   16  1046     241633 byte  1589    1199220 byte  outbound-full-relay    4faketorv2pbfu7x.onion:8333
+   19  577      181679 byte  390      148951 byte  outbound-full-relay    kfake4vctorjv2o2.onion:8333
+   20  11         1248 byte  13         1283 byte  block-relay-only       [2600:fake:64d9:b10c:4436:aaaa:fe:bb]:8333
+   21  11         1248 byte  13         1299 byte  block-relay-only       XX.XXX.X.155:8333
+   22  5           103 byte  1           102 byte  feeler                 XX.XX.XXX.173:8333
+   23  11         1248 byte  12         1255 byte  block-relay-only       XX.XXX.XXX.220:8333
+   24  3           103 byte  1           102 byte  feeler                 XXX.XXX.XXX.64:8333
+…
```
Comment (azuchi) [inline contrib/tracing/README.md:104]: `TYPE` value different from the implementation.

File: doc/tracing.md [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+
+#### Tracepoint `net:inbound_message`
+
+Is called when a message is received from a peer over the P2P network. Passes
+information about our peer, the connection and the message as arguments.
+
+Arguments passed:
+1. Peer ID as `int64`
+2. Peer Address and Port (IPv4, IPv6, Tor v3, I2P, ...) as `pointer to C-style String` (max. length 68 characters)
+3. Connection Type (inbound, feeler, outbound-full-relay, ...) as `pointer to C-style String` (max. length 20 characters)
```
Comment (azuchi) [inline doc/tracing.md:66]: `TYPE` value different from the implementation in this document.

**Review by azuchi — CHANGES_REQUESTED** [review]

File: Dockerfile [hunk 13]
```diff
@@ -1,15 +1,27 @@
...
+ENV BUILD_PACKAGES "build-essential libtool autotools-dev automake pkg-config bsdmainutils curl git ca-certificates ccache g++ make cmake gdb gdbserver rsync zip openssh-server lldb ssh python3 bison"
+
+ENV TRACE_PACKAGES "bpfcc-tools systemtap-sdt-dev python3-bpfcc"
+
+ENV DEBIAN_FRONTEND=noninteractive
+RUN apt-get update
+RUN apt-get install --no-install-recommends --no-upgrade -qq $PACKAGES
+RUN apt-get install --no-install-recommends --no-upgrade -qq $BUILD_PACKAGES
+RUN apt-get install --no-install-recommends --no-upgrade -qq $TRACE_PACKAGES
+RUN apt-get install --no-install-recommends --no-upgrade -qq linux-headers-generic
```
Comment (azuchi) [inline Dockerfile:16]: The installation of these packages was done by tapyrus-builder, so it is not necessary or they should be installed with tapyrus-builder.

**Review by azuchi — CHANGES_REQUESTED** [review]

File: Dockerfile [hunk 14]
```diff
@@ -28,5 +29,26 @@ COPY entrypoint.sh /usr/local/bin/
...
+ENV BPFTRACE_VERSION v0.15.0
+WORKDIR /bpftrace
+RUN wget https://github.com/iovisor/bpftrace/releases/download/${BPFTRACE_VERSION}/bpftrace && \
+    chmod +x bpftrace && \
+    mv bpftrace /bin && \
+    mkdir -p /tracing && \
+    wget https://github.com/iovisor/bpftrace/archive/${BPFTRACE_VERSION}.zip && \
+    unzip ${BPFTRACE_VERSION}.zip && \
+    cp -r bpftrace*/tools /tracing &&\
+    mv /kernel/usr/src/linux-headers* /kernel/usr/src/linux-headers
```
Comment (azuchi) [inline Dockerfile:47]: Why not just install bpftrace with apt in tapyrus/builder?

File: test/config copy.ini [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+# These environment variables are set by the build process and read by
+# test/functional/test_runner.py and test/util/tapyrus-util-test.py
+
+[environment]
+SRCDIR=/Users/navia/Documents/Projects/Tapyrus/tapyrus-core-nb/FunctionalTestTimeout
+BUILDDIR=/Users/navia/Documents/Projects/Tapyrus/tapyrus-core-nb/FunctionalTestTimeout/build
+EXEEXT=
+RPCAUTH=/Users/navia/Documents/Projects/Tapyrus/tapyrus-core-nb/FunctionalTestTimeout/share/rpcauth/rpcauth.py
```
Comment (azuchi) [inline test/config copy.ini:12]: Individual environmental information is undesirable.

File: Dockerfile [hunk 16]
```diff
@@ -9,14 +10,25 @@ COPY . .
...
+
+RUN apt-get update
+RUN apt-get install --no-install-recommends --fix-missing --no-upgrade -qq $TRACE_PACKAGES
+
+WORKDIR /kernel
+COPY  --from=ksrc /kernel-dev.tar .
+RUN tar xf kernel-dev.tar
+RUN mv /kernel/usr/src/linux-headers* /kernel/usr/src/linux-headers
+ENV BPFTRACE_KERNEL_SOURCE=/kernel/usr/src/linux-headers
+
```
Comment (azuchi) [inline Dockerfile:31]: I appreciate your work, but but from this PR, let's not apply USDT to container.

**Review by azuchi — CHANGES_REQUESTED** [review]

File: test/config copy.ini (hunk 15 above)
Comment (Naviabheeman) [reply test/config copy.ini]: this file was added by accident. it shouldn't be there

File: Dockerfile (hunk 14 above)
Comment (Naviabheeman) [reply Dockerfile]: done.

**Review by azuchi — APPROVED** [review]
